File: src/kirala_peasy_ffi.erl
-module(kirala_peasy_ffi).
-compile(export_all).

start() ->
    inets:start(),
    ssl:start().

to_char_list(Str) when is_list(Str) ->
    Str;
to_char_list(Str) when is_binary(Str) ->
    unicode:characters_to_list(Str).

readline(Prompt) ->
    io:get_line(Prompt).

show(X) when is_integer(X) ->
    integer_to_binary(X);
show(X) when is_float(X) ->
    float_to_binary(X);
show(X) when is_atom(X) ->
    atom_to_binary(X);
show(X) when is_list(X) ->
    gleam@string:join( lists:map(fun show/1, X), <<" ">> );
show(X) when is_tuple(X) ->
    gleam@string:join( lists:map(fun show/1, tuple_to_list(X)), <<" ">> );
show(X) when is_binary(X) ->
    X.


println(Any) ->
    io:format("~ts~n", [show(Any)]).

print(Any) ->
    io:format("~ts", [show(Any)]).

read_file(Filepath) ->
    file:read_file(Filepath).

read_text_file(Filepath) ->
    read_file(Filepath).

http_get(Url,Headers) ->
    {ok, {ResStatus, ResHeaders, ResBody}} = httpc:request(get,{Url, Headers}),
    {ResStatus, ResHeaders, ResBody}.

now() ->
    timestamp( erlang:now() ).

timestamp({Mega, Secs, Micro}) ->
    (Mega*1000*1000*1000*1000 + Secs * 1000 * 1000 + Micro) / 1000000.0.

get_env(Key,DefaultValue) ->
    os:getenv(Key,to_char_list(DefaultValue)).

get_env_int(Key,DefaultValue) ->
    os:getenv(Key,integer_to_list(DefaultValue)).

File: src/kirala_peasy_ffi.mjs
import fs from "fs";

/*
import * as readline_lib from "readline";


const rl = readline_lib.createInterface({
    input: process.stdin,
    output: process.stdout
});

export function readline_async(p,callback) {
    let answer = null
    rl.question(p, (a) => {
        callback(a);
       // rl.close();
    });
}
*/

import * as readline_sync from "readline-sync";

export function readline(prompt) {
    return readline_sync.question(prompt);
}

export function show(val) {
    if( val === null ) {
        return "null";
    }
    else if( val === undefined ) {
        return "undefined";
    }
    else {
        return val.toString();
    }
}

let _outputs = "";

export function println(val) {
    _outputs += val;
    console.log(_outputs);
    _outputs = "";
}

export function print(val) {
    _outputs += val;
}

export function read_file(filepath) {
    let raw = fs.readFileSync(filepath);
    return raw;
}

export function read_text_file(filepath) {
    let text = fs.readFileSync(filepath, "utf-8");
    return text;
}

export function http_get_text(url, headers) {
}

export function now() {
    return Date.now();
}


export function get_env(key,default_value) {
    let val = process.env[key];
    if( val === undefined || val == null ) {
        return default_value;
    }
    else {
        return val;
    }
}

export function get_env_int(key,default_value) {
    let val = process.env[key];
    if( val === undefined || val == null ) {
        return default_value;
    }
    else {
        return parseInt(val);
    }
}
